### PR TITLE
incompressible: recover ice-slurry transport data, fix PCL viscosity (100x), add MITSW freezing curve, stabilise json output

### LINKED ISSUE #3303: IceEA/IceNA/IcePG conductivity and viscosity cannot be regenerated — the source CSVs are latin-1 and their loads are commented out
Found while reviewing the rebase of #3295. Filing separately: the fix is a data/pipeline change, and both #3294 and #3295 deliberately leave it alone.

## What

`json/IceEA.json`, `json/IceNA.json` and `json/IcePG.json` each ship fitted `conductivity` and `viscosity` blocks that the current pipeline **cannot reproduce**. Re-running the fitter clears both properties to `notdefined`.

The committed coefficients are not garbage — they have plausible fit residuals, so they were produced by a real fit at some point:

| fluid | property | type | shape | NRMS |
|---|---|---|---|---|
| IceEA | conductivity | polynomial | 4x6 | 1.63e-3 |
| IceEA | viscosity | exppolynomial | 4x6 | 2.60e-3 |
| IceNA | conductivity | polynomial | 4x6 | 5.13e-5 |
| IceNA | viscosity | exppolynomial | 4x6 | 5.26e-5 |
| IcePG | conductivity | polynomial | 4x6 | 1.05e-3 |
| IcePG | viscosity | exppolynomial | 4x6 | 2.13e-4 |

## Why the refit clears them

`SecCoolIceData` (`CPIncomp/SecCoolFluids.py:413`) loads only `Hfusion`. The conductivity and viscosity loads immediately below are commented out:

```python
self.specific_heat.xData, self.specific_heat.yData, self.specific_heat.data = self.getArray(dataID='Hfusion')
self.specific_heat.source = self.specific_heat.SOURCE_DATA

#self.conductivity.xData,self.conductivity.yData,self.conductivity.data   = self.getArray(dataID='Cond')
#self.conductivity.source = self.conductivity.SOURCE_DATA

#self.viscosity.xData,self.viscosity.yData,self.viscosity.data   = self.getArray(dataID='Mu')
#self.viscosity.source = self.viscosity.SOURCE_DATA
```

With no data loaded, the fit produces an all-zero conductivity and a `None` viscosity, and `clearUnfittedCoefficients` correctly marks both `notdefined`.

## The likely cause, and why the fix is small

**The source data is still in the repository.** All six files exist:

```
CPIncomp/data/SecCool/xTables/.../Ice{EA,NA,PG}_{Cond,Mu}.csv
```

Every one of them is **latin-1 encoded** (`·` in the unit headers) and raises `UnicodeDecodeError` on a default UTF-8 read. That is almost certainly why the loads were commented out, and it matches finding 5 in `DATA_AUDIT.md` (added in #3295), which flags these as orphaned latin-1 files and notes that anyone reviving them must pass `encoding="latin-1"`.

So the fix is probably: read those CSVs with `encoding="latin-1"`, uncomment the two load blocks, refit, and confirm the regenerated coefficients agree with what is committed. If they agree, the data is recovered and reproducible. If they do not, the committed values came from something else and should be cleared to `notdefined` instead.

## Why no guard catches this

This is the same class as #1331 / #2567 — coefficients in `json/` that do not correspond to a real, reproducible fit — but it is invisible to every check added in #3294 and #3295:

- not all-zero, so `test_no_all_zero_fits_shipped` passes
- not an optimizer starting guess, so `test_no_placeholder_guesses_shipped` passes
- structurally valid and finite, so the loader-contract checks pass
- the golden-master refit test in #3295 pins only DowJ, LiBr and AKF, so these three are not compared

A test that refits **every** fluid and asserts each committed property is reproducible would catch it, but that is a much slower check than the current golden master and needs its own design.

## Impact

Low but not zero. The values load and evaluate today, so `INCOMP::IceEA` conductivity and viscosity return numbers. Those numbers are unverifiable against the source data as the pipeline currently stands, and would silently vanish the next time anyone regenerates `json/`.

## Related

- #2567 reports `T_freeze` failing for exactly these three fluids across their whole composition range — the unaddressed part of that issue. Whatever is wrong with `SecCoolIceData`'s data loading may well cover both.
- #3302 — the 24 all-zero conversion blocks, a sibling data-provenance problem in the same corpus.
- #3294, #3295 — the guards that catch the neighbouring cases but not this one.

### LINKED ISSUE #3384: INCOMP::PCL (Paracryol) viscosity is a factor of 100 too high
`json/PCL.json` ships viscosity coefficients that are **100x above** what the fitting pipeline produces from PCL's own source table. Anyone who has called `INCOMP::PCL` for viscosity has had a value two orders of magnitude too large.

Found while building a check that every committed fit still matches the data it was fitted to (the same class of problem as #3303 and #2488, caught by a different mechanism).

## What is wrong

The fluid is declared with `viscosityFactor=1e-5`:

```python
SecCoolSolutionData(sFile='Paracryol', sFolder='xPure', name='PCL',
                    desc='Paracryol, Aliphatic Hydrocarbon',
                    ref='Sulzer1999,Skovrup2013', ..., viscosityFactor=1e-5)
```

but the committed coefficients correspond to `1e-3`. The fit is `exponential`, i.e. `exp(A/(T + B) - C)`, and only `C` differs:

| | A | B | C |
|---|---|---|---|
| committed | 797.6255 | -92.39176 | **5.714384** |
| refitted from the data | 797.6255 | -92.39176 | **10.31955** |

`exp(10.31955 - 5.714384) = 100.0` exactly. `A` and `B` are bit-identical, so this is purely a scale factor that was changed in the code at some point without `json/` being regenerated.

## Which one is correct: `1e-5`

Three independent lines, strongest first.

**1. Prandtl number from PCL's own other tables.** No external source needed, since density, heat capacity and conductivity for the same fluid are committed alongside:

| T [degC] | rho | cp [J/kg/K] | k [W/m/K] | Pr at 1e-5 | Pr at 1e-3 |
|---|---|---|---|---|---|
| -40 | 820.4 | 1750 | 0.152 | 120 | 12016 |
| 0 | 792.0 | 1910 | 0.149 | 34 | 3371 |
| 180 | 664.2 | 2632 | 0.136 | **5.6** | **558** |

A light hydrocarbon at 180 degC with density 664 kg/m3 and conductivity 0.136 W/m/K cannot have Pr = 558.

**2. Kinematic viscosity.** At 0 degC: 3.3 mm2/s with `1e-5`, 332 mm2/s with `1e-3`. The latter is gear-oil territory and is flatly contradicted by the same fluid's density and conductivity.

**3. A peer fluid.** Paracryol (Sulzer AG) is listed among heat transfer fluids alongside Methylcyclopentane, Syltherm XLT, Marlotherm X and Therminol D-12. D-12 runs 14.8 mm2/s at -50 degC, 1.23 at 40 degC and 0.65 at 100 degC:

| T [degC] | Paracryol at 1e-5 | Paracryol at 1e-3 | Therminol D-12 |
|---|---|---|---|
| -40 | 12.7 | 1272 | 14.8 (at -50) |
| 40 | 1.62 | 162 | 1.23 |
| 100 | 0.80 | 80 | 0.65 |

`1e-5` puts Paracryol right where a light hydrocarbon belongs. `1e-3` is off by two orders of magnitude at every temperature.

## Why nothing caught it

The guards added in #3294 and #3295 all pass on this file, and so does the golden-master refit test in #3295:

- the coefficients are not all-zero, not a placeholder guess, structurally valid and finite
- the golden master compares a **refit against `json/`**, but both sides of that comparison come from the same fitter run, so it cannot see that the committed file was produced under a different scale factor
- `NRMS` is computed against the loaded (correctly scaled) data, so the committed residual looks healthy

What catches it is comparing the **committed coefficients** against the **raw source grid**. On that measure PCL viscosity is off by 103x; the worst legitimate deviation anywhere else in the corpus is 33% (VMA viscosity, a steep low-temperature curve on a low-order polynomial), and the median is 0.08%.

## Fix

- regenerate `json/PCL.json` so the viscosity constant becomes 10.31955
- add `test_committed_fits_still_describe_their_source_data`, which evaluates every committed fit against the grid its fluid loads and fails above 100% deviation

Worth checking whether any other `*Factor` argument has drifted the same way since its fluid was last regenerated. The new test covers the corpus as it stands, so a second case would have shown up; but a factor changed in future without regenerating would now be caught at test time rather than shipped.

### PR DESCRIPTION
**Pull request by jowr**
### Description of the Change

This started as a fix for #3303 and grew to cover several related problems in the incompressible-fluid pipeline. Most are the same family: coefficients committed in `json/` that no longer correspond to the data or the code that produced them.

**1. Ice-slurry conductivity and viscosity were silently lost (#3303)**

`Ice{EA,NA,PG}_{Cond,Mu}.csv` each contain exactly one latin-1 byte, the 0xB7 middle dot in the unit header (`W/(m.K)`, `Pa.s`). `skiprows=3` discards that header line, but numpy decodes the whole file before skipping, so the read raises `UnicodeDecodeError` anyway. `SecCoolSolutionData.__init__` performs that read inside a bare `try/except`, so the failure was swallowed: both properties silently dropped out of every refit and were written to `json/` as `notdefined`.

Re-encoding the six files to plain ASCII recovers the data. The refit reproduces the committed coefficients to ~3e-7 relative, same fit type and same shape, which confirms the shipped values did come from exactly these tables. So `json/` needed no correction for these three fluids, only the ability to reproduce it.

Worth noting for reviewers: uncommenting the loads in `SecCoolIceData` is not what fixed this. The base class was already reading both properties; the encoding was the whole problem. The reads are repeated in the subclass, outside the bare `except`, and go through `getRequiredArray`, which raises when a grid is missing or unreadable. That second part matters: `getArray` only raises for a file that exists and cannot be decoded, and returns `(None, None, None)` for a file it cannot find, which would have produced the same silent `notdefined` by a different route.

**2. PCL (Paracryol) viscosity was a factor of 100 too high (#3384)**

Found by the new source-data test described below. `json/PCL.json` shipped viscosity coefficients corresponding to `viscosityFactor=1e-3`, while `SecCoolFluids.py` has declared `1e-5` for some time. The fit is `exp(A/(T + B) - C)`; `A` and `B` are bit-identical between the committed and refitted versions and only `C` differs, by `ln(100)` to within the 7-digit rounding (`exp(10.31955 - 5.714384)` = 99.9996).

`1e-5` is the correct one:

- Prandtl number computed from PCL's own raw density, heat capacity and conductivity tables is 5.6 at 180 degC with `1e-5` and 558 with `1e-3`. (Using the fitted coefficients rather than the raw tables gives 5.8 and 584; the point is the order of magnitude either way.) A light hydrocarbon with density 664 kg/m3 and conductivity 0.136 W/m/K cannot have Pr in the hundreds.
- Kinematic viscosity at 0 degC is 3.3 mm2/s with `1e-5`, 332 mm2/s with `1e-3`. The latter is gear-oil territory and contradicts the same fluid's other properties.
- Therminol D-12, listed alongside Paracryol as a peer heat transfer fluid, runs 1.23 mm2/s at 40 degC and 0.65 at 100 degC. Paracryol at `1e-5` gives 1.62 and 0.80. At `1e-3` it would give 162 and 80. (D-12's published low-temperature point is 14.8 mm2/s at -50 degC; PCL's range stops at -40 degC, where it gives 12.7, so those two are close but not measured at the same temperature.)

`json/PCL.json` is regenerated. This is a user-visible change in returned values.

**3. MITSW now has a freezing temperature, from IAPWS-08 (#2567)**

#3294 removed the placeholder that had `T_freeze` returning ~0 K for MIT Seawater, so the property currently raises. This restores it from a real source, which is what fwitte asked for in the issue.

Sharqawy 2010 supplies every other MITSW property but publishes no freezing temperature, and the MIT seawater library has no such function among its 24 properties, so there is no same-source option. IAPWS-08 is the right one.

A **cubic** in `(x - xbase)`, not the quintic proposed in the thread, and the reason is worth stating. The true freezing point of pure water at 1 atm is 273.1525 K, because 273.15 K (0 degC) is the air-saturated ice point. IAPWS-08 returns the true value, so at zero salinity the real curve sits 2.5 mK **above** this fluid's own `Tmin` of 273.15 K. `checkT` throws when `T < T_freeze`, so a curve crossing `Tmin` would start rejecting `PropsSI` calls at exactly `T = Tmin` with zero salinity, which work today. A quintic lands 0.7 mK above `Tmin` there; the cubic smooths that corner and stays ~7 mK below it across the whole range, still true after the writer rounds to 7 significant digits.

The cost is small and stated: the shipped curve reads ~10 mK low at zero salinity and is within 10 mK of IAPWS-08 elsewhere.

This does **not** widen the usable range. `T_freeze` is used in exactly one place, an extra lower bound in `checkT`, and it is below `Tmin` everywhere, so no validity behaviour changes. What it buys is that `trivial_keyed_output(iT_freeze)` returns a real number instead of raising. Lowering `Tmin` to meet `T_freeze`, floated upthread in #2567, would push the Sharqawy correlations outside their 0-120 degC validity, so `Tmin` is untouched.

`LiBr` is deliberately left `notdefined`: below the eutectic it forms ice and above it LiBr hydrates crystallise, so a single polynomial over 0-75% would misrepresent both branches. fwitte reached the same conclusion.

**4. Generated json is now rounded to 7 significant digits**

`json.dumps` writes all 17 decimal digits of a double. The committed files carry 7, so re-running the pipeline on a different numpy/scipy rewrote 116 of the 127 files with last-digit noise, and a real change could not be told apart from the churn. That churn is a large part of why #3303 sat unnoticed. The rounding helpers already existed (`SolutionData.round`) but were never called from the writer.

Rounding is applied once, to the whole fluid dict just before `json.dumps`, so it covers coefficients, NRMS, `Tbase`/`xbase`/`TminPsat` and the Chebyshev blocks under one rule. `allow_nan=False` is also set, so a non-finite value stops the pipeline instead of being written as a bare `NaN` token that is not valid JSON, and `writeFluidList` now raises on a failed write instead of logging and reporting success.

**5. Published composition limits were outside the real range (#2567)**

The fluid tables formatted `x_min`/`x_max` with two decimals. MAM2 stops at `x = 0.236` and the table printed `0.24`, so the documented upper limit is a composition the backend rejects with `Your composition 0.24 is not between 0.078 and 0.236`. Simulating a sweep between the formatted limits reproduces exactly the four mass-based fluids reported in #2567 (MAM2, VCA, VKC, VMG) and no volume-based fluid, matching the report. Every limit in `json/` is exact at three decimals, so widening the format fixes it and rounds nothing away.

`Incompressibles.rst` now also documents `ifraction_min`/`ifraction_max`, which is how to read the enforced limits at runtime. The absence of any signpost for that API is what led the reporter to scrape the docs table in the first place.

**6. Error messages now name the fluid and the property (#2567)**

Asking for a property a fluid does not define produced `IncompressibleFluid.cpp (388): The function type is not specified ("[0]"), are you sure the coefficients have been set?` with no fluid, no property, and no indication of whether this was a data gap or a bug. It now reads:

```
INCOMP::IceEA does not define a freezing temperature: no fit for this property is
shipped in its fluid data, so there is nothing to evaluate. This is a gap in the
reference data for this fluid, not a problem with your inputs.
```

All twelve `INCOMPRESSIBLE_NOT_SET` sites go through one helper. It returns the exception rather than throwing it, so every call site keeps its own `throw` and the compiler still sees each switch case as terminating; a helper that threw internally would need `[[noreturn]]` for that.

This also fixes a copy-paste slip in `T_rho`: it switches on `density.type`, but both of its diagnostics reported `specific_heat.type`, naming the wrong property.

Finally, `DATA_AUDIT.md` and the user docs record why `IceEA`/`IceNA`/`IcePG` define no freezing temperature and should not: the composition axis of an ice slurry is the ice mass fraction, so the fluid is already at solid-liquid equilibrium everywhere in its range, and the equilibrium temperature is the temperature axis of the tables rather than a separate curve over `x`.

### Benefits

- `INCOMP::PCL` viscosity returns correct values. It has been 100x too high.
- `INCOMP::MITSW` freezing temperature returns a real IAPWS-08 value instead of raising.
- `IceEA`/`IceNA`/`IcePG` conductivity and viscosity are reproducible from source again. Before this, the next person to regenerate `json/` would have silently deleted them.
- Regenerating `json/` produces a diff that means something. Two full regenerations from a cleared hash cache are byte-identical, so a real coefficient change stands out instead of hiding among 5000 lines of formatting noise.
- The composition limits in the online tables can be fed back into `set_mass_fractions()` without hitting a `ValueError`, and there is a documented way to read the enforced limits at runtime.
- A user who asks for an undefined property learns which fluid, which property, and that the cause is missing reference data rather than their own inputs.
- A new class of bug is caught in CI: a committed fit that no longer describes the data it was fitted from. That class covers #1331, #2488, #3303 and #3384, and no existing guard detected it.

### Possible Drawbacks

- **`INCOMP::PCL` viscosity changes by 100x.** This is a correction, but anyone who compensated for the old value downstream will see different numbers.
- **MITSW's freeze curve is deliberately ~10 mK low at zero salinity**, to keep it clear of `Tmin` so `checkT` behaviour is unchanged. A reviewer who wants IAPWS agreement at that one corner instead of the `Tmin` margin should say so; the two cannot both be had with `Tmin` where it is.
- **`json/` is still not fully regenerated here.** Only `PCL.json` and `MITSW.json` are. A full regeneration rewrites ~120 other files, mostly to round the Chebyshev blocks for the first time. That diff is real content rather than formatting noise, but it deserves its own review rather than riding along here.
- **A missing or unreadable ice-slurry data file now raises during fluid construction**, and a failed json write now aborts the generator, instead of both being swallowed. That is the point, but both are behaviour changes.
- **The 100% threshold in `test_committed_fits_still_describe_their_source_data` is empirical.** It sits three times above the worst legitimate deviation in the corpus (33%, VMA viscosity) and a hundred times below the PCL failure.
- The published composition limits in the docs tables gain a decimal place, a cosmetic change to generated documentation.

### Verification Process

**Python**, all run locally: 162 tests in `dev/incompressible_liquids/` (87 passed, 75 skipped; 27 tests existed before this PR). The skips are the pure-fluid cases of the composition-range test, which assert nothing and are reported as skips rather than counted as passes.

Every new guard was checked against the pre-fix state, to be sure it can actually fail rather than merely pass:

- restoring the latin-1 CSVs fails `test_loaded_data_files_are_ascii`, the three ice refit tests, and errors out the SecCool fixture
- renaming `IceEA_Cond.csv` fails `test_missing_ice_data_file_raises_instead_of_silently_dropping`
- reverting `json/PCL.json` fails the source-data test with `PCL.viscosity: committed fit is off its own data by 103x`
- reverting the `x()` formatter to two decimals fails the composition-range test on exactly MAM2, VCA, VKC and VMG
- a swallowing `writeFluidList` and a fail-fast one both fail `test_write_fluid_list_raises_when_a_fluid_cannot_be_serialised`
- for MITSW's freeze curve, three wrong versions were tried: perturbed coefficients fail both new tests; `T_freeze` removed fails the IAPWS test; and **the quintic passes the IAPWS test, because it fits better, while failing the `Tmin` test**. That last one is the evidence the two tests are not redundant.

Other checks:

- The rounding was measured, not assumed. Effect on evaluated properties across every fluid's own T,x range: median 6.5e-8, worst 9.5e-5 (PCR saturation pressure, where the exponential amplifies). The fits' own residuals reach 4.8%.
- Regenerating MITSW leaves every one of its existing properties unchanged to within 1.2e-8 relative; only `T_freeze` is new.
- Two full pipeline regenerations from a cleared hash cache produce byte-identical output, and all 127 files parse as strict JSON.
- `./dev/ci/preflight.sh` passes its `incomp-sanity` and `clang-format` gates.

**An adversarial review of this branch** (required by `CLAUDE.md` before opening a PR, and run late here) found that three guards added earlier in the branch could pass while checking nothing: the "fails loudly" claim held only for an undecodable file and not a missing one; the ASCII test could decode zero files and still go green; and the composition-range test would collapse into a single skip on an empty glob, with a green exit code. It also found the test's `INCOMP_EPSILON` was 4.5e7 times looser than the backend's, and a docstring still quoting a C++ error string this PR deletes. CodeRabbit then found a fourth, in the fix for those: `allow_nan=False` made `toJSON` able to raise into a caller that swallowed it. All are fixed, each verified by making it fail first.

**C++:** the session that wrote this could not compile it. The container cannot configure the build because CPM fetches valijson and the other dependencies from github.com and the egress policy returns 403, so `cmake` never completes. In place of a compiler, before pushing: the novel construct (a const member returning an exception by value, thrown inside a switch) was compiled and run standalone under `g++ -std=c++17 -Wall -Wextra -Werror`; every one of the twelve `case` labels was checked to still exist with its own throw directly beneath it; braces and parentheses were checked to balance; and `clang-format` 18.1.8 reported no new violations.

CI has since built it on Linux, macOS and Windows with no failures, and `clang-tidy`, `include-what-you-use`, `cppcheck` and the installed-header hygiene gate all pass on the changed files. A `[INCOMP]` Catch2 run on a real build is still the right last check before merge, particularly for the PCL and MITSW value changes.

### Applicable Issues

Closes #3303
Closes #3384

Relates to #2567, and now addresses most of it: the MAM2/VCA/VKC/VMG limits, the ice-slurry behaviour and documentation, the error message, and a real `T_freeze` for MITSW. Still not closed, for two reasons. The `MMG2` case from the original report could not be reproduced (its current limits, 0.0 to 0.205, format inward and its fit evaluates cleanly across the whole range), and `LiBr` still has no freezing temperature because its crystallisation behaviour changes branch at the eutectic.

## Summary by CodeRabbit

- **Documentation**
  - Clarified composition limits and querying options for incompressible mixtures.
  - Documented slurry-ice composition behavior and unsupported freezing-temperature queries.

- **Bug Fixes**
  - Restored conductivity and viscosity data for ice-slurry fluids.
  - Missing or unreadable property data now reports clear errors.
  - Improved messages for unsupported fluid properties.

- **Improvements**
  - Added freezing-temperature data for MIT seawater.
  - Updated PCL property correlations and standardized numeric precision.
  - Improved validation of fluid-property data and fitted results.
  - Fluid data generation now reports failed outputs instead of completing silently.

### CODE REVIEW

**Comment by jowr** [comment]
> 🟠 Major · Propagate strict JSON serialization failures from batch generation. · `WriterObjects.py:585-590`

Verified and fixed in e32decf. The finding is correct, and it is a hole I opened one commit earlier.

`allow_nan=False` (added in 805a3f5) made `toJSON` able to raise for the first time in practice. `writeFluidList` caught `TypeError`/`ValueError`, printed it, and continued, then printed `" ... done"` and returned normally, so `all_incompressibles.py` reported `All done, bye` and exited 0 while `json/<name>.json` was left missing or holding the previous run's numbers. That is the same shape as the bug this whole PR is about, a stale coefficient file that nothing complains about, reintroduced at the last step by the commit that was closing fail-open holes. The old docstring even stated the hazard outright (*"a clean return does NOT mean every file was written"*) and nobody, me included, acted on it.

Fixed by aggregating rather than re-raising immediately, so one bad fluid does not mask the state of the rest, with the collected failures raised after the loop. A normal return now means every file really was written.

`test_write_fluid_list_raises_when_a_fluid_cannot_be_serialised` pins both halves: that the batch raises, and that the healthy fluid was still written. It fails when the raise is removed. A full regeneration of all 127 files still completes, so nothing in the corpus depends on the old behaviour.

Note for reviewers, since this is a behaviour change beyond what the PR body currently lists: the generator now aborts the run on a write failure instead of skipping the fluid.

On the docstring-coverage check: I added docstrings to the five functions this PR introduces that were missing them. The remainder of the 44 it counted are pre-existing functions in the files this PR happens to touch (`fitFluid`, `getFile`, `get_hash`, the table helpers, and so on); documenting all of `WriterObjects.py` is a separate job and would bury this diff.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: dev/incompressible_liquids/test_writer_output.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    file while the pipeline claims success is exactly the failure this branch
+    exists to remove, so the batch writer has to surface it.
+    """
+    writer = SolutionDataWriter()
+
+    class FakeFluid(object):
+        def __init__(self, name):
+            self.name = name
+
+    fluids = [FakeFluid("GoodOne"), FakeFluid("BadOne")]
```
[the remark this answers is not in the document]
Comment (jowr) [reply dev/incompressible_liquids/test_writer_output.py]: Adopted in 77f98e0, though your stated reasoning does not quite hold for this code, and the real problem is slightly worse than described.

> The current order allows a fail-fast `writeFluidList` to raise on `BadOne` after writing `GoodOne`, so the test still passes

It does not, as written. I made `writeFluidList` fail-fast and ran it: the test fails, because `pytest.raises(match="BadOne")` does not match a re-raised serialization error, whose text is `Out of range float values are not JSON compliant` and never names the fluid.

But that is luck, not design. The test was relying on the wording of an exception it does not control. A serialization error that happened to name the offending field, which is perfectly plausible, would have satisfied the regex, and the test would have been pinning "it raises" while looking like it pinned "it keeps going". So the fix is right for a different reason than the one given.

Reordering makes the distinction structural rather than incidental: with the failure first, only an implementation that continues past it can reach `GoodOne` at all, so the `written` assertion carries the proof instead of the regex. I also gave that assertion a message saying what it is really testing.

Verified against three implementations:

| implementation | test |
|---|---|
| silently swallows (the original bug) | fails |
| fail-fast, message names the fluid (used to slip through) | fails |
| aggregates and raises at the end (current) | passes |

Worth noting for anyone reading the history: this is the third round on this branch where a check I added could pass without doing its job. Same root cause every time, which is writing the assertion for the behaviour I intended instead of asking which wrong implementations still satisfy it. The bot catches are appreciated.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: dev/incompressible_liquids/test_writer_output.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    the test that stops someone "improving" the fit and breaking it.
+    """
+    with open(os.path.join(JSON_DIR, "MITSW.json")) as fh:
+        fluid = json.load(fh)
+    entry = fluid["T_freeze"]
+    if entry.get("coeffs") in (None, "null"):
+        pytest.skip("MITSW ships no T_freeze")
+
+    xs = np.linspace(fluid["xmin"], fluid["xmax"], 241)
+    values = np.array([_evaluate_T_freeze(entry, x, fluid["xbase"]) for x in xs])
```
[the remark this answers is not in the document]
Comment (jowr) [reply dev/incompressible_liquids/test_writer_output.py]: Correct, and fixed in 33b0c10. I built the counterexample rather than taking it on trust, and it is worse than "could in principle".

A parabola peaked exactly halfway between two grid points:

```
T(x) = (Tmin + 1) - 1.7e7 * (x - xbase - u0)^2,   u0 midway between samples 100 and 101
```

scores **273.0875 K** on the 241-point grid, so the old test passed it, while actually reaching **274.15 K** — a full kelvin above `Tmin`, which is exactly the failure the test exists to prevent.

Adopted the endpoints-plus-derivative-roots approach. A polynomial attains its extremes on an interval only at an endpoint or a critical point, so that set is exhaustive for *any* polynomial, not just the cubic shipped today. That matters here because the test's purpose is to constrain coefficients nobody has written yet.

Verified against four cases:

| coefficients | result |
|---|---|
| adversarial parabola (passed the old grid) | **fails** |
| the quintic from #2567 | **fails**, as before |
| degree-1 curve | **fails** by assertion on the endpoint, not a crash on the empty derivative-root list |
| the shipped cubic | passes |

I kept `np.atleast_1d` around the root array for the degree-0 case, which is the one shape where numpy's `.roots()` behaviour is easiest to get wrong.

For the record: this is the fourth time on this branch that a check I wrote could pass while the property it advertised was violated. Same root cause every time, which is validating the case in front of me instead of asking which inputs still satisfy the assertion. A sampling grid should have been an obvious instance, given that this very file already turns on that question.

---
_Generated by [Claude Code](https://claude.ai/code)_
